Update documentation for v2 release (#276)
kubernetes-csi/csi-proxy: docs/USAGE.md

> # Usage

Author: It's a bit long and has a lot of nuanced details re. containerd and HostProcess containers. I think it might be easier to keep this in a separate doc and point to it. I like to keep READMEs a hub of pointers to relevant information.

}

func (mounter *CSIProxyMounterV1) PathExists(path string) (bool, error) {
    isExistsResponse, err := mounter.FsClient.PathExists(context.Background(),
        // the request type is exposed in pkg/filesystem
        &fs.PathExistsRequest{
            Path: mount.NormalizeWindowsPath(path),
        })
    if err != nil {
        return false, err
    }
    return isExistsResponse.Exists, err
}
```

### Deployment

CSI driver containers need to run as HostProcess pods for the CSI Proxy commands to complete privileged operations.
This can be done by updating the driver pod spec, typically embedded in a Daemonset.
The Kubernetes Windows HostProcess Pod [docs](https://kubernetes.io/docs/tasks/configure-pod-container/create-hostprocess-pod/) goes into more detail about each field.

```yaml
spec:
  securityContext:
    hostNetwork: true
    windowsOptions:
      hostProcess: true
      runAsUserName: "NT AUTHORITY\\SYSTEM"
```

Using HostProcess pods have a few important consequences on the deployed containers, as noted in the HostProcess docs.
- HostProcess containers have no file system or resource isolation, so they have a complete view of the host machine’s file system.

> This means that paths passed in **arguments need to be complete host machine paths**.

Reviewer: ```suggestion
Paths in the PodSpec must be absolute paths with respect to the host.
```

On the other hand, depending on the containerd and the Windows OS version, Kubernetes volume mounts are either mounted relative to a subdirectory of the host machine specified by an environment variable `%CONTAINER_SANDBOX_MOUNT_POINT%` or **mounted relative to the host process root**. See [HostProcess Caveats](#hostprocess-caveats).
- HostProcess pods can only contain HostProcess containers.
Often, the CSI node registrar is deployed in the same pod as the driver, so file paths for both containers need to be updated.

> - Named pipes and Unix domain sockets are not supported and should be accessed via their path on the host.

Reviewer: ```suggestion
- Named pipes and Unix domain sockets are not supported and should be accessed via their absolute path with respect to the host.
```

Instead of mounting host process paths such as `c:\var\lib\kubelet` to each container’s own filesystem, the container can directly access these paths.

If drivers are migrating from CSI Proxy v1, note that since the current version (v2) no longer has a separate binary running on the host, drivers no longer need to mount named pipes for each API group/version used. Related volumes and volume mounts can be safely deleted.

Here is an example driver's deployment configuration.

```yaml
kind: DaemonSet
apiVersion: apps/v1
metadata:
  name: csi-storage-node-win
spec:
  selector:
    matchLabels:
      app: csi-driver-win
  template:
    metadata:
      labels:
        app: csi-driver-win
    spec:
      serviceAccountName: csi-node-sa
      tolerations:
      - key: "node.kubernetes.io/os"
        operator: "Equal"
        value: "win1809"
        effect: "NoSchedule"
      nodeSelector:
        kubernetes.io/os: windows
      containers:
        - name: csi-driver-registrar
          image: k8s.gcr.io/sig-storage/csi-node-driver-registrar:v2.1.0
          args:
            - --v=5

>             - --csi-address=unix://c:\var\lib\kubelet\plugins\pd.csi.storage.gke.io\csi.sock

Reviewer: let's use placeholders for the CSI Driver instead of using PD CSI Driver as an example

>             - --csi-address=unix://c:\var\lib\kubelet\plugins\pd.csi.storage.gke.io\csi.sock

Reviewer: Are we going to use `CONTAINER_SANDBOX_MOUNT_POINT`?

Author: It's best to not do so. The container files are mounted at `CONTAINER_SANDBOX_MOUNT_POINT` across versions, but volume mounts are either mounted relative to `CONTAINER_SANDBOX_MOUNT_POINT` in containerd v1.6 or the host FS root (which critically isn't the same as `CONTAINER_SANDBOX_MOUNT_POINT`) in containerd v1.7. I'd just stick with absolute paths to ensure portability between versions.

Reviewer: sounds good
